pandas/core/sorting.py

```diff
@@ -162,6 +162,10 @@ def decons_obs_group_ids(comp_ids, obs_ids, shape, labels, xnull):
     xnull: boolean,
         if nulls are excluded; i.e. -1 labels are passed through
     """
+    labels = list(labels)
```

Comment:
this non-performant, can we not add a special case path

Reply:
@jreback  So I suggested a [new approach](https://github.com/pandas-dev/pandas/pull/28634#discussion_r339307374). it doesn't need labels : )

Reply:
and this one is likely non-performant

Reply:
?? @jreback did you read this?

> @jreback I think a modification should be under the method of `BinGrouper` because this kind of issue occurs only with `groupby` having `freq`. And if it has `freq`, then it would be `BinGrouper`. I could somewhat manage to do it in `decons_obs_group_ids` though it's quite tricky as you see. In the first place, if the type of grouper is BinGrouper, then it doesn't need both `labels` and `decons_obs_group_ids` and It's ok to just return unique comp_ids(checked by `pytest pandas` but not absolutely sure).
> 
> However why unique comp_ids? that's because `rep`, a partial function which is used at making result index expects unique ids that would be result index. `rep` seems to do its job well. And we know we should change something about `self.recons_labels`.
> 
> If you don't like current PR(tricky length checking), consider the overriding again, please. Thank you

> My suggestion is this:
> 
> ```diff
> diff --git a/pandas/core/groupby/ops.py b/pandas/core/groupby/ops.py
> @@ -792,6 +792,11 @@ class BinGrouper(BaseGrouper):
> +    @cache_readonly
> +    def recons_labels(self):
> +        comp_ids, _, _ = self.group_info
> +        return [np.unique(comp_ids)]
> +
> diff --git a/pandas/core/sorting.py b/pandas/core/sorting.py
> @@ -162,10 +162,6 @@ def decons_obs_group_ids(comp_ids, obs_ids, shape, labels, xnull):
> -    labels = list(labels)
> -    unique_comp_ids = np.unique(comp_ids)
> -    if (shape[0] != len(unique_comp_ids)) and (shape[0] == len(labels[0])):
> -        return [unique_comp_ids]
> ```
> 
> I confirmed it passes all tests. If you don't mind, I'll push it.

I think it couldn't be slower than the original.

Reply:
can you run appropriate asvs and test this assumption, list-ifying is generally note a great idea here.

Reply:
@jreback  Did you mean the assumption `labels = list(labels)`, by any chance? or [adding independent recons_labels method under the BinGrouper](https://github.com/pandas-dev/pandas/pull/28634#discussion_r339307374)? I couldn't get it why did you mention `list-ifying`. I agree that  `labels = list(labels)` can slower performance when `xnull` is `True`. But what I want to suggest was adding independent recons_labels method, which is dismissed approach before.

I'm running asv with _recons_labels under the Bingrouper_ for now, as it takes a lot of time.

Reply:
you were right... np.unique is not enough. Therefore I'm trying this `[np.nonzero(np.r_[self.bins, 0] != np.r_[0, self.bins])[0][:-1]]` which is inspired by here:
https://github.com/pandas-dev/pandas/blob/ec5335645ac683f7a04127f78f36e3996929999f/pandas/core/groupby/generic.py#L649-L653

Reply:
I solved performance issue.